common/channel/src/libra_channel.rs

```diff
@@ -0,0 +1,148 @@
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::Waker;
+
+use futures::async_await::FusedStream;
+use futures::stream::Stream;
+use futures::task::Context;
+use futures::Poll;
+
+/// ValidatorMessage trait needs to be implemented by every message which gets pushed into the Libra
+/// channel so that we can ensure fairness among validators
+pub trait ValidatorMessage {
+    /// Extract the Validator from which this message arrived. This
+    /// will be used for ensuring fairness among validators.
+    fn get_validator(&self) -> &str;
+}
+
+/// LibraChannelType is an enum which can be used as a configuration option for
+/// LibraChannel
+pub enum LibraChannelType {
+    LIFO,
+    FIFO,
+}
+
+struct PerValidatorQueue<T> {
+    queue_type: LibraChannelType,
+    per_validator_queue: HashMap<String, VecDeque<T>>,
+    max_queue_size: usize,
+}
+
+impl<T: ValidatorMessage> PerValidatorQueue<T> {
+    fn new(queue_type: LibraChannelType, max_queue_size: usize) -> Self {
+        if max_queue_size == 0 {
+            panic!("max_queue_size should be > 0");
+        }
+        Self {
+            queue_type,
+            per_validator_queue: HashMap::new(),
+            max_queue_size,
+        }
+    }
+
+    fn add_message(&mut self, message: T) {
+        let validator = message.get_validator();
+        if !self.per_validator_queue.contains_key(validator) {
+            self.per_validator_queue
+                .insert(validator.to_string(), VecDeque::new());
+        }
+        if let Some(q) = self.per_validator_queue.get_mut(validator) {
+            q.push_back(message);
+            if q.len() == self.max_queue_size + 1 {
+                match self.queue_type {
+                    // Drop old message for FIFO
+                    LibraChannelType::FIFO => q.pop_front(),
+                    // Drop new message for LIFO
+                    LibraChannelType::LIFO => q.pop_back(),
+                };
+            }
+        }
+    }
+
+    // TODO: Come up with a better data structure to efficiently
+    // do a round robin among validators
+    fn get_message(&mut self) -> Option<T> {
+        for q in self.per_validator_queue.values_mut() {
+            if !q.is_empty() {
+                return match self.queue_type {
+                    LibraChannelType::FIFO => q.pop_back(),
+                    LibraChannelType::LIFO => q.pop_front(),
+                };
+            }
+        }
+        None
+    }
+}
+
+struct SharedState<T> {
+    /// The data stored by this LibraChannel
+    /// TODO This needs to be updated to a data structure which can do round robin for validators
+    data: PerValidatorQueue<T>,
+
+    /// The thread can use this after calling `put` to tell
+    /// `LibraChannelReceiver`'s task to wake up, and process the next item
+    waker: Option<Waker>,
+}
+
+/// The sending end of LibraChannel
+#[derive(Clone)]
+pub struct LibraChannelSender<T: ValidatorMessage> {
```

Comment:
Since the crate is named "libra-channel" you don't need to prefix every type in this crate with "LibraChannel". I would recommend dropping the prefix throughout the crate.

Reply:
Done